cpp/include/kvikio/utils.hpp

```diff
@@ -52,36 +53,84 @@ inline constexpr std::size_t page_size = 4096;
- * isn't the case, a CUfileException is thrown.
- *
- * @return The current cuda context
+ * @param ptr Device pointer to query
+ * @return The device ordinal
  */
-[[nodiscard]] inline CUcontext get_current_context(const void* check_owning_devPtr = nullptr)
+[[nodiscard]] inline int get_device_ordinal_from_pointer(const void* devPtr)
 {
-  if (check_owning_devPtr != nullptr) {
-    CUdeviceptr current_ctx_devPtr{};
-    CUdeviceptr dev_ptr = convert_void2deviceptr(check_owning_devPtr);
-
-    CUresult const err = cudaAPI::instance().PointerGetAttribute(
-      &current_ctx_devPtr, CU_POINTER_ATTRIBUTE_DEVICE_POINTER, dev_ptr);
-    if (err != CUDA_SUCCESS || current_ctx_devPtr != dev_ptr) {
-      throw CUfileException("The current CUDA context must own the given device memory");
+  CUdeviceptr dev_ptr = convert_void2deviceptr(devPtr);
+  int ret;
+  CUDA_DRIVER_TRY(
+    cudaAPI::instance().PointerGetAttribute(&ret, CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL, dev_ptr));
+  return ret;
+}
+
+/**
+ * @brief RAII wrapper for a CUDA primary context
+ */
+class CudaPrimaryContext {
+ public:
+  CUdevice dev;
+  CUcontext ctx;
+
+  CudaPrimaryContext(int device_ordinal)
+  {
+    CUDA_DRIVER_TRY(cudaAPI::instance().DeviceGet(&dev, device_ordinal));
+    CUDA_DRIVER_TRY(cudaAPI::instance().DevicePrimaryCtxRetain(&ctx, dev));
+  }
+  CudaPrimaryContext(const CudaPrimaryContext&) = delete;
+  CudaPrimaryContext& operator=(CudaPrimaryContext const&) = delete;
+  CudaPrimaryContext(CudaPrimaryContext&&)                 = delete;
+  CudaPrimaryContext&& operator=(CudaPrimaryContext&&) = delete;
+  ~CudaPrimaryContext()
+  {
+    try {
+      CUDA_DRIVER_TRY(cudaAPI::instance().DevicePrimaryCtxRelease(dev), CUfileException);
+    } catch (const CUfileException& e) {
+      std::cerr << e.what() << std::endl;
     }
   }
+};
 
-  CUcontext ret{};
-  CUDA_DRIVER_TRY(cudaAPI::instance().CtxGetCurrent(&ret));
-  return ret;
+/**
+ * @brief Return a CUDA that can be used with the given device pointer
+ *
+ * @param devPtr Device pointer to query
+ * @return Usable CUDA context
+ */
+[[nodiscard]] inline CUcontext get_context_from_pointer(const void* devPtr)
+{
+  static std::map<int, CudaPrimaryContext> _primary_contexts;
+  CUdeviceptr dev_ptr = convert_void2deviceptr(devPtr);
+
+  // First we check if a context has been associated with `devPtr`.
+  // Notice, this is not the case for stream ordered device memory allocations.
+  // See <https://docs.nvidia.com/cuda/cuda-c-programming-guide/#pointer-attributes>
+  CUcontext ctx;
+  const CUresult err =
+    cudaAPI::instance().PointerGetAttribute(&ctx, CU_POINTER_ATTRIBUTE_CONTEXT, dev_ptr);
+  if (err != CUDA_ERROR_INVALID_VALUE) {
+    // If this isn't the case, we grab the current context
+    CUDA_DRIVER_TRY(cudaAPI::instance().CtxGetCurrent(&ctx));
+  } else {
+    CUDA_DRIVER_TRY(err);  // Check for errors from previous asynchronous launches
+  }
```

Comment:
I think this logic is wrong? Let's try and work through it:

The options for the allocation are:
1. stream ordered: `CU_POINTER_ATTRIBUTE_CONTEXT` will return `CUDA_SUCCESS` and put `ctx = nullptr`
2. not stream ordered: `CU_POINTER_ATTRIBUTE_CONTEXT` will return `CUDA_SUCCESS` and put `ctx = some_context`
3. not a UVM allocation: `CU_POINTER_ATTRIBUTE_CONTEXT` will return `CUDA_ERROR_INVALID_VALUE` 
4. Some other thing: some other cuda error

So if we hit case 3 or 4, we want to error, if we hit case 2, we want to return the context that the query returned, if we hit case 1 we need to do some more work: I think we can just use `CU_POINTER_ATTRIBUTE_DEVICE_POINTER` query and then do something based on that (either return a non-null current context or else build a primary context and return that).

So perhaps:
```c++
CUcontext ctx;
// either the pointer is valid (CUDA_SUCCESS), or it isn't (some other error)
CUDA_DRIVER_TRY(cudaAPI::instance().PointerGetAttribute(&ctx, CU_POINTER_ATTRIBUTE_CONTEXT, dev_ptr));
if (ctx != nullptr) {
    // not a stream-ordered allocation, this context is fine.
    return ctx;
}
// stream-ordered allocation, let's see if the current device context exists and we can use it
CUDA_DRIVER_TRY(cudaAPI::instance().CtxGetCurrent(&ctx));
if (ctx != nullptr) {
    CUdeviceptr target{};
    auto const err = cudaAPI::instance().PointerGetAttribute(&target, CU_POINTER_ATTRIBUTE_DEVICE_POINTER, dev_ptr);
    if (err == CUDA_SUCCESS && target == dev_ptr) return ctx;
}
// Either no context, or the current context can't access the device pointer?
int ord = get_device_ordinal_from_pointer(devPtr);
_primary_contexts.try_emplate(ord, ord);
return _primary_contexts.at(ord).ctx;
```

WDYT?